lib/blemulator/internal/simulation_manager.dart

```diff
@@ -23,18 +30,79 @@ class SimulationManager {
   Future<void> _destroyClient() async {}
 
   Future<void> _startDeviceScan() async {
-    _peripherals.forEach((peripheral) {
-      _scanSubscriptions.add(peripheral
-          .onScan(allowDuplicates: true)
-          .listen((scanResult) {
-            print(scanResult);
-            return _bridge.addScanResult(scanResult);
-          }));
+    _peripherals.values.forEach((peripheral) {
+      _scanSubscriptions
+          .add(peripheral.onScan(allowDuplicates: true).listen((scanResult) {
+        print(scanResult);
+        return _bridge.publishScanResult(scanResult);
+      }));
     });
   }
 
   Future<void> _stopDeviceScan() async {
     _scanSubscriptions.forEach((subscription) => subscription.cancel());
     _scanSubscriptions.clear();
   }
+
+  Future<void> _connectToDevice(String identifier) async {
+    if (_peripherals[identifier] != null) {
+      _connectionStateSubscriptions.putIfAbsent(
+          identifier,
+              () => _peripherals[identifier]
+              .connectionStateStream
+              .listen((connectionState) {
+            _bridge.publishConnectionState(
+                _peripherals[identifier], connectionState);
+
+            if (connectionState == PeripheralConnectionState.disconnected) {
+              _connectionStateSubscriptions[identifier].cancel();
+              _connectionStateSubscriptions.remove(identifier);
+            }
+          }));
+
+      bool canConnect = await _peripherals[identifier].onConnectRequest();
+
+      if (canConnect) {
+        _peripherals[identifier].onConnect();
+      } else {
+        return Future.error(<String, dynamic>{
+          "errorCode": BleErrorCode.DeviceAlreadyConnected,
```

Comment:
I think that this information should be gotten from peripheral, because now we hardcoded just a one reason why we cannot connect with selected peripheral.

Reply:
Maybe onConnectRequest should throw an error? Or maybe onConnect should determine how the connection process should look like and then we could just return _peripherals[identifier].onConnect from here?

Reply:
1) I made a mistake with conditions, it's fixed now.
2) Ad `onConnectRequest()` I think this gives user more control over the behaviour of the simulated peripheral. I think there are some devices that would not even attempt connection and would just reject any such requests.